File: sdk/js/examples/real_tx_smoke.js
// Phase 1 Step 3 smoke — real signed Transfer end-to-end.
//
// Steps:
//   1. Load testnet/keys/validator_0.key (genesis-funded sender).
//   2. Read sender's current nonce + slot from the API.
//   3. Build a signed Transfer to a random recipient.
//   4. Submit via SDK — expect HTTP 202 (`accepted: true`).
//   5. Subscribe to /v1/events and wait for a `BatchExecuted` frame.
//   6. Re-read sender + recipient via /v1/account; assert nonce + balances moved.
//
// Exits non-zero on any step that fails.

import { readFileSync } from 'node:fs';
import { randomBytes } from 'node:crypto';

import {
  SubtickClient,
  buildSignedTransfer,
  derivePublicKey,
  subscribeEvents,
} from '../src/index.js';

const baseUrl = process.argv[2] || 'http://127.0.0.1:18080';
const keyPath = process.argv[3] || '../../subtick/testnet/keys/validator_0.key';

console.log(`[smoke] base=${baseUrl} key=${keyPath}`);

// 1. Load the sender's key (hex-encoded 32-byte seed).
const seedHex = readFileSync(keyPath, 'utf8').trim();
const seed = Buffer.from(seedHex, 'hex');
if (seed.length !== 32) {
  console.error(`[smoke] expected 32-byte hex seed, got ${seed.length} bytes`);
  process.exit(2);
}
const senderPub = derivePublicKey(seed);
const senderHex = senderPub.toString('hex');
console.log(`[smoke] sender pubkey: ${senderHex}`);

const client = new SubtickClient({ baseUrl });

// 2. Health + sender state.
const health = await client.health();
console.log(`[smoke] api ok — height=${health.height} slot=${health.slot}`);

const senderBefore = await client.getAccount(senderHex);
console.log(`[smoke] sender before — balance=${senderBefore.balance} nonce=${senderBefore.nonce}`);

// 3. Build the tx.
const recipient = randomBytes(32);
const recipientHex = recipient.toString('hex');
const amount = 100n;
const ttl = BigInt(health.slot) + 10_000n;

const txHex = buildSignedTransfer({
  privateKey: seed,
  senderPubkey: senderPub,
  recipient,
  amount,
  nonce: senderBefore.nonce,
  ttl,
});
console.log(
  `[smoke] tx built — recipient=${recipientHex.slice(0, 12)}... amount=${amount} ` +
    `nonce=${senderBefore.nonce} ttl=${ttl} hex_len=${txHex.length}`,
);

// 4. Subscribe BEFORE submit so we don't miss the executed frame.
//    Wait for a frame that ACTUALLY applied a tx (group_applied_txs > 0) —
//    the orderer fires empty BatchExecuted frames every round when there's
//    nothing pending, so we have to filter past those to find ours.
let executedFrame = null;
let allFrames = 0;
const sub = subscribeEvents(
  baseUrl,
  (event) => {
    if (event.type !== 'BatchExecuted') return;
    allFrames += 1;
    console.log(
      `[ws] frame#${allFrames} shard=${event.shard_id} ` +
        `applied=${event.group_applied_txs} ` +
        `rej_nonce=${event.group_rejected_nonce} ` +
        `rej_bal=${event.group_rejected_balance} ` +
        `batch=${event.batch_id.slice(0, 12)}...`,
    );
    if (executedFrame === null && event.group_applied_txs > 0) {
      executedFrame = event;
    }
  },
  {
    onOpen: () => console.log('[ws] connected'),
    onError: (err) => console.log(`[ws] error: ${err.message}`),
  },
);

// Give the WS upgrade ~250ms to settle.
await sleep(250);

// 5. Submit.
let submit;
try {
  submit = await client.sendTx(txHex);
  console.log(`[smoke] sendTx accepted: ${submit.txHash}`);
} catch (err) {
  console.error(`[smoke] sendTx FAILED: ${err.message} (status=${err.status})`);
  sub.close();
  process.exit(3);
}
if (!submit.accepted) {
  console.error(`[smoke] expected accepted=true, got ${JSON.stringify(submit)}`);
  sub.close();
  process.exit(4);
}

// 6. Wait up to 10 s for a BatchExecuted frame.
const frameDeadline = Date.now() + 10_000;
while (executedFrame === null && Date.now() < frameDeadline) {
  await sleep(100);
}
if (executedFrame === null) {
  console.error('[smoke] no BatchExecuted frame received within 10s — block path may be off');
  sub.close();
  process.exit(5);
}

// 7. Re-read state.
await sleep(500); // let state apply settle
const senderAfter = await client.getAccount(senderHex);
console.log(`[smoke] sender after — balance=${senderAfter.balance} nonce=${senderAfter.nonce}`);

let recipientAfter;
try {
  recipientAfter = await client.getAccount(recipientHex);
  console.log(
    `[smoke] recipient after — balance=${recipientAfter.balance} nonce=${recipientAfter.nonce}`,
  );
} catch (err) {
  console.log(`[smoke] recipient lookup: ${err.message}`);
  recipientAfter = { balance: '0', nonce: 0 };
}

// 8. Assertions.
const senderBeforeBal = BigInt(senderBefore.balance);
const senderAfterBal = BigInt(senderAfter.balance);
const recipientAfterBal = BigInt(recipientAfter.balance);

const balanceDelta = senderBeforeBal - senderAfterBal;
const nonceDelta = senderAfter.nonce - senderBefore.nonce;

let pass = 0;
let fail = 0;
function check(name, cond, detail) {
  if (cond) {
    pass += 1;
    console.log(`PASS ${name}`);
  } else {
    fail += 1;
    console.log(`FAIL ${name} :: ${detail}`);
  }
}

// Note (v0 executor — apply_batch_to_delta in dag/state_apply.rs):
// The Phase 6 v1.5 executor's apply phase deducts the transferred AMOUNT
// only, not the fee. Fee burn / validator reward is out of scope until the
// executor's economic finalisation step lands. So the sender balance delta
// equals exactly `amount`, not `amount + fee`.
check('nonce incremented by 1', nonceDelta === 1, `delta=${nonceDelta}`);
check(
  'sender balance dropped by amount (v0 executor: no fee burn yet)',
  balanceDelta === amount,
  `delta=${balanceDelta} expected=${amount}`,
);
check('recipient credited by amount', recipientAfterBal === amount, `recipient=${recipientAfterBal}`);
check('BatchExecuted applied=1 frame received', executedFrame !== null, '<no frame>');

sub.close();
console.log(`\nresult: ${pass} pass, ${fail} fail`);
process.exit(fail === 0 ? 0 : 1);

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

File: sdk/python/examples/real_tx_smoke.py
"""Phase 1 Step 3 smoke for the Python SDK — mirror of the JS smoke."""

from __future__ import annotations

import os
import secrets
import sys
import time
from pathlib import Path

# Make the package importable without installing.
sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from subtick_sdk import (
    AccountNotFound,
    SubtickClient,
    build_signed_transfer,
    derive_public_key,
    subscribe_events,
)


def main() -> int:
    base_url = sys.argv[1] if len(sys.argv) > 1 else "http://127.0.0.1:18080"
    key_path = (
        sys.argv[2]
        if len(sys.argv) > 2
        else "../../subtick/testnet_real_smoke/keys/v0.key"
    )
    print(f"[smoke-py] base={base_url} key={key_path}")

    seed_hex = Path(key_path).read_text().strip()
    seed = bytes.fromhex(seed_hex)
    if len(seed) != 32:
        print(f"[smoke-py] expected 32-byte hex seed, got {len(seed)} bytes")
        return 2

    sender_pub = derive_public_key(seed)
    sender_hex = sender_pub.hex()
    print(f"[smoke-py] sender pubkey: {sender_hex}")

    client = SubtickClient(base_url=base_url)
    health = client.health()
    print(f"[smoke-py] api ok - height={health.height} slot={health.slot}")

    sender_before = client.get_account(sender_hex)
    print(f"[smoke-py] sender before - balance={sender_before.balance} nonce={sender_before.nonce}")

    recipient = secrets.token_bytes(32)
    recipient_hex = recipient.hex()
    amount = 100
    ttl = health.slot + 10_000

    tx_hex = build_signed_transfer(
        private_key=seed,
        sender_pubkey=sender_pub,
        recipient=recipient,
        amount=amount,
        nonce=sender_before.nonce,
        ttl=ttl,
    )
    print(
        f"[smoke-py] tx built - recipient={recipient_hex[:12]}... "
        f"amount={amount} nonce={sender_before.nonce} ttl={ttl} hex_len={len(tx_hex)}"
    )

    executed_frame: dict | None = None

    def on_event(event: dict) -> None:
        nonlocal executed_frame
        if event.get("type") != "BatchExecuted":
            return
        applied = int(event.get("group_applied_txs", 0))
        print(
            f"[ws] shard={event.get('shard_id')} applied={applied} "
            f"rej_nonce={event.get('group_rejected_nonce')} "
            f"rej_bal={event.get('group_rejected_balance')} "
            f"batch={str(event.get('batch_id', ''))[:12]}..."
        )
        if executed_frame is None and applied > 0:
            executed_frame = event

    sub = subscribe_events(base_url, on_event)
    time.sleep(0.25)

    try:
        submit = client.send_tx(tx_hex)
        print(f"[smoke-py] send_tx accepted: {submit.tx_hash}")
    except Exception as err:  # noqa: BLE001
        print(f"[smoke-py] send_tx FAILED: {err}")
        sub.stop()
        return 3

    if not submit.accepted:
        print(f"[smoke-py] expected accepted=True, got {submit}")
        sub.stop()
        return 4

    deadline = time.time() + 10.0
    while executed_frame is None and time.time() < deadline:
        time.sleep(0.1)

    if executed_frame is None:
        print("[smoke-py] no BatchExecuted (applied>0) frame in 10s")
        sub.stop()
        return 5

    time.sleep(0.5)  # let state apply settle
    sender_after = client.get_account(sender_hex)
    print(f"[smoke-py] sender after  - balance={sender_after.balance} nonce={sender_after.nonce}")

    try:
        recipient_after = client.get_account(recipient_hex)
        print(
            f"[smoke-py] recipient after - balance={recipient_after.balance} nonce={recipient_after.nonce}"
        )
        recipient_balance = recipient_after.balance
    except AccountNotFound:
        print("[smoke-py] recipient lookup: account not found")
        recipient_balance = 0

    balance_delta = sender_before.balance - sender_after.balance
    nonce_delta = sender_after.nonce - sender_before.nonce

    pass_count = 0
    fail_count = 0

    def check(name: str, ok: bool, detail: str = "") -> None:
        nonlocal pass_count, fail_count
        if ok:
            pass_count += 1
            print(f"PASS {name}")
        else:
            fail_count += 1
            print(f"FAIL {name} :: {detail}")

    check("nonce incremented by 1", nonce_delta == 1, f"delta={nonce_delta}")
    check(
        "sender balance dropped by amount (v0 executor: no fee burn yet)",
        balance_delta == amount,
        f"delta={balance_delta} expected={amount}",
    )
    check(
        "recipient credited by amount",
        recipient_balance == amount,
        f"recipient={recipient_balance}",
    )
    check(
        "BatchExecuted applied>=1 frame received",
        executed_frame is not None,
        "<no frame>",
    )

    sub.stop()
    print(f"\nresult: {pass_count} pass, {fail_count} fail")
    return 0 if fail_count == 0 else 1


if __name__ == "__main__":
    raise SystemExit(main())
